tools/bin/check_images_exist.sh

```diff
@@ -1,78 +1,133 @@
-    curl --silent -H "Authorization: Bearer $token" -f -lSL "$URL" > /dev/null
+    echo -e "$blue_text""\n\n\tURL: $URL""$default_text"
+    curl -H "Authorization: Bearer $token" --location --silent --show-error --dump-header header.txt "$URL" > /dev/null
+    curl_success=$?
   else
     URL=https://hub.docker.com/v2/repositories/"$1"/tags/"$2"
-    printf "\tURL: %s\n" "$URL"
-    curl --silent -f -lSL "$URL" > /dev/null
+    echo -e "$blue_text""\n\n\tURL: $URL""$default_text"
+    curl --silent --show-error --location --dump-header header.txt "$URL" > /dev/null
+    curl_success=$?
+    # some bullshit to get the number out of a header that looks like this
+    # < content-length: 1039
+    # < x-ratelimit-limit: 180
+    # < x-ratelimit-reset: 1665683196
+    # < x-ratelimit-remaining: 180
+    docker_rate_limit_remaining=$(grep 'x-ratelimit-remaining: ' < header.txt | grep --only-matching --extended-regexp "\d+")
+    # too noisy when set to < 1.  Dockerhub starts complaining somewhere around 10
+    if test "$docker_rate_limit_remaining" -lt 20; then
+      echo -e "$red_text""We are close to a sensitive dockerhub rate limit!""$default_text"
+      echo -e "$red_text""SLEEPING 60s sad times""$default_text"
+      sleep 60
+      docker_tag_exists $1 $2
+    elif test $docker_rate_limit_remaining -lt 50; then
+      echo -e "$red_text""Rate limit reported as $docker_rate_limit_remaining""$default_text"
+    fi
+  fi
+  if test $curl_success -ne 0; then
+    echo -e "$red_text""Curl Said this didn't work.  Please investigate""$default_text"
+    exit 1
   fi
 }
 
 checkPlatformImages() {
-  echo "Checking platform images exist..."
-  docker-compose pull || exit 1
-  echo "Success! All platform images exist!"
+  echo -e "$blue_text""Checking platform images exist...""$default_text"
+  #Pull without printing progress information and send error stream because that where image names are
+  docker-compose pull --quiet 2> compose_output
+  docker_compose_success=$?
+  images_pulled_count=$(docker images | wc -l)
+  if test $images_pulled_count -eq 0; then
+    echo -e "$red_text""Nothing was pulled!  This script may be broken! We expect to pull something""$default_text"
+    exit 1
+  elif test $docker_compose_success -eq 0; then
+    echo -e "$blue_text""Docker successfully pulled all images""$default_text"
+  else
+    echo -e "$red_text""docker-compose failed to pull all images""$default_text"
+    cat compose_output
+    exit 1
+  fi
 }
 
 checkNormalizationImages() {
+  echo -e "$blue_text""Checking Normalization images exist...""$default_text"
   # the only way to know what version of normalization the platform is using is looking in NormalizationRunnerFactory.
   local image_version;
   factory_path=airbyte-commons-worker/src/main/java/io/airbyte/workers/normalization/NormalizationRunnerFactory.java
-  # This check should fail if the file doesn't exist, so just try to ls the file
-  ls $factory_path > /dev/null
```

Comment:
I wanted to write more clear error output here... and then I got carried away

Reply:
this is glorious